src/repo.rs

```diff
@@ -0,0 +1,618 @@
+            lender: Address,
+            loan_token: Address,
+            security_token: Address,
+            loan_amount: U256,
+            security_amount: U256,
+            interest_rate: U256,
+            activation_deadline: u64,
+            return_deadline: u64) {
+
+            self.storage.write(&BORROWER_KEY, &H256::from(borrower).into());
+            self.storage.write(&LENDER_KEY, &H256::from(lender).into());
+            self.storage.write(&LOAN_TOKEN_KEY, &H256::from(loan_token).into());
+            self.storage.write(&SECURITY_TOKEN_KEY, &H256::from(security_token).into());
+            self.storage.write(&LOAN_AMOUNT_KEY, &loan_amount.into());
+            self.storage.write(&SECURITY_AMOUNT_KEY, &security_amount.into());
+            self.storage.write(&INTEREST_RATE_KEY, &interest_rate.into());
+            self.storage.write(&ACTIVATION_DEADLINE_KEY, &U256::from(activation_deadline).into());
+            self.storage.write(&RETURN_DEADLINE_KEY, &U256::from(return_deadline).into());
+        }
+
+        // Tries to activate contract
+        fn accept(&mut self) -> bool {
+            let sender = ext::sender();
+            if self.is_active() {
+                panic!("Cannot accept, contract has activated already");
+            }
+            if ext::timestamp() > self.activation_deadline() {
+                ext::suicide(&sender);
+            }
+
+            let lender_address = self.lender_address();
+            let borrower_address = self.borrower_address();
+
+            // Accept by borrower
+            if sender == borrower_address {
+                self.storage.write(&BORROW_ACCEPTED_KEY, &U256::from(1).into());
+            }
+            // Accept by lender
+            else if sender == lender_address {
+                self.storage.write(&LEND_ACCEPTED_KEY, &U256::from(1).into());
+            } else {
+                panic!("Only for participants");
+            }
+
+            // Wait for all parties to accept
+            if !(self.borrower_acceptance() && self.lender_acceptance()) {
+                return false;
+            }
+
+            let mut loan_token = Token::new(self.loan_token_address());
+            let mut security_token = Token::new(self.security_token_address());
+            let loan_amount = self.loan_amount();
+            let security_amount = self.security_amount();
+
+            let this_contract_address = ext::address();
+            // Transfer security from borrower_address to the contract address
+            assert!(security_token.transferFrom(borrower_address, this_contract_address, security_amount));
+            // Transfer loan to the borrower address
+            assert!(loan_token.transferFrom(lender_address, borrower_address, loan_amount));
+
+            return true;
+        }
+
+        fn terminate(&mut self) -> bool {
+            if !self.is_active() {
+                panic!("Can't terminate contract: contract hasn't activated");
```

Comment:
how to end it if other parity has no desire to activate it?

Reply:
anybody can do in through accept call.. But anyway, I'll add it to terminate too